docs/proposals/drafts/20211007-pluggable-authenticators.md

```diff
@@ -0,0 +1,146 @@
+3. Oauth2 Providers
+
+## 1. Importing for side-effects
+
+Like [Option 3](#3-oauth2-providers) except that we do not have a static list of supported authenticators. Instead, each authenticator can register themselves through a map of authenticator types and their factories.
+
+To onboard a new authenticator, the following is needed:
+
+1. Implement the corresponding interface.
+2. Implement a factory.
+3. Register the factory in the authenticators' factories map, this is the part where it registers itself.
+
+The factory should receive three parameters: a config map of strings and interfaces, tenant's name, and a logger instance.
+
+With this approach, an example of tenant configuration would like:
+
+```
+- name: test-oidc
+  id: 1610b0c3-c509-4592-a256-a1871353dbfa
+  authenticator:
+    type: oidc
+    config:
+      clientID: test
+      clientSecret: xyz
+      issuerCAPath: ./tmp/certs/ca.pem
+      issuerURL: http://127.0.0.1:5556/dex
+      redirectURL: https://localhost:8443/oidc/test-oidc/callback
+      usernameClaim: email
+  rateLimits:
+    - endpoint: "/api/metrics/v1/.+/api/v1/receive"
+      limit: 100
+      window: 1s
+    - endpoint: "/api/logs/v1/.*"
+      limit: 100
+      window: 1s
+```
+
+### 2. Go Plugin Module
+
+In this approach, a generic interface for Authentication providers (Authenticators) is introduced. Authentication providers then implement that interface in a go plugin. This plugin then will be built, and the compiled plugin is loaded at run time by Observatorium API, how and where to load the plugin module is defined by the tenants’ config file.
+
+This approach is easy to implement, very flexible in terms of adding more Authenticators without any additional implementation and integration complexity. However, there are few concerns when it comes to trust and security, since Observatorium API would consume a pre-built plugin binary as an authenticator.
+
+**Reference**: [Writing Modular Go Programs with Plugins](https://medium.com/learning-the-go-programming-language/writing-modular-go-programs-with-plugins-ec46381ee1a9)
+
+## 3. Oauth2 Providers
+
+Similar to the [generic Oauth2 providers in grafana](https://github.com/grafana/grafana/blob/main/pkg/login/social/social.go), Oauth2 providers can be introduced and injected into the supported Oauth2 besides to the current implementation of OIDC and mTLS to prevent any disruption for the current adopters. While this approach provides flexibility and it considers to be an elegant and clean way of introducing more Authenticators support, it requires a lot of changes in the Observatorium API.
+
+However, introducing a new authenticator with private or custom implementation would require changing/patching existing Go source files in the Observatorium API to achieve that. As a result, adopters need to re-apply their patch whenever migrating to a new Observatorium API release with risk to be affected by merge conflicts that must be resolved manually.
```

Comment:
```suggestion
However, introducing a new authenticator with a private or custom implementation would require changing/patching existing Go source files in the Observatorium API to achieve that. As a result, adopters need to re-apply their patch whenever migrating to a new Observatorium API release with the risk of being affected by merge conflicts that must be resolved manually.
```

Reply:
`Doing X would require Y to achieve that` sounds redundant to me. I know that I need `Y` to achieve `X` because I said that `X` requires `Y.
```suggestion
However, introducing a new authenticator with private or custom implementation would require changing/patching existing Go source files in the Observatorium API. As a result, adopters need to re-apply their patch whenever migrating to a new Observatorium API release with risk to be affected by merge conflicts that must be resolved manually.
```